core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -531,6 +591,38 @@ class ZkUtils(val zkClient: ZkClient,
     dataAndStat
   }
 
+  /**
+    * Callback for asynchronous version of readDataMaybeNull
+    */
+  private class readDataMaybeNullCallback extends DataCallback {
+    def processResult(rc: Int,
+                      path: String,
+                      ctx: Object,
+                      readData: Array[Byte],
+                      stat: Stat) {
+      val (path, topCallback, topCtx) = ctx.asInstanceOf[(String, ((Option[String], Stat), scala.Any) => Unit, scala.Any)]
+      Code.get(rc) match {
+        case Code.OK =>
+          val dataStr : Option[String] = Some(ZKStringSerializer.deserialize(readData).asInstanceOf[String])
+          topCallback((dataStr, stat), topCtx)
+        case Code.NONODE | Code.CONNECTIONLOSS | Code.SESSIONEXPIRED=>
+          info("The node at %s has gone away while reading it, ".format(path))
+          readDataMaybeNullAsync(path, topCallback, topCtx)
```

Comment:
This reads a bit weird to me, if the session expired, then resubmitting the operation doesn't make much sense. Is this because ZkClient is supposed to create a new session under the hood?